### PR TITLE
Addons Engine + Marketplace

### PR DESCRIPTION
**Pull request by melkodary**
#315 

### PR:
Three major components created:
- **marketplace_manager**: is a server for hosting available addons.
- **addons_manager**: is a server, providing manipulation to addons database. User decide whether to run an addon or disable it by modifying the state of the addon in the database.
- **addons_monitor**: Polls the addons_manager and docker to achieve the target state of the addon. It is responsible for running, stopping and retrying addons. _Addons monitor utilizes docker outside docker: basically allows a container to manage containers of the host (outside) by mounting the socket volume. This approach is more safe/secure in_ comparison to dind.

#### Simple Run:
1. inside root_orchestrator: `docker compose up --build --force-recreate addons_monitor -d`
2. make a post request to `[POST] localhost:11102/api/v1/marketplace/addons`:

```
    {
      "name": "testing-addon",
      "services": [
        {
          "command": "sh -c 'while true; do echo \"Hello, World (testing!!) !\"; sleep 10; done'",
          "image": "alpine",
          "service_name": "my-alpine"
        }
   ]}
```

5. make another api request to `[POST] localhost:11101/api/v1/addons`:

```
    {
        "marketplace_id": (id from the response of the above request)
    }
```

### CODE REVIEW

File: addons_marketplace/marketplace-manager/api/v1/marketplace_blueprint.py
Comment (Malyuk-A) [inline addons_marketplace/marketplace-manager/api/v1/marketplace_blueprint.py:22]: Is this method intended to add/post multiple addons at once or only single ones? If only single once then I think this method would be better placed at the singular "MarketPlaceAddonController" below not at this plural one.

File: addons_marketplace/marketplace-manager/api/v1/schema.py
Comment (Malyuk-A) [inline addons_marketplace/marketplace-manager/api/v1/schema.py:25]: What exactly are services in the context of marketplace addons?

File: addons_marketplace/marketplace-manager/services/marketplace_service.py
Comment (Malyuk-A) [inline addons_marketplace/marketplace-manager/services/marketplace_service.py:13]: Do I understand this correctly?:
The verification is checking if the image can be pulled?

File: addons_marketplace/marketplace-manager/services/marketplace_service.py
Comment (Malyuk-A) [inline addons_marketplace/marketplace-manager/services/marketplace_service.py:32]: What reason is there to store addons in the DB that failed the verification? Is is due to the long asynch behavior and letting callers wait for the response?
(I.e. why not just store addons in the DB that passed verification?)

File: addons_marketplace/marketplace-manager/api/v1/marketplace_blueprint.py
Comment (Malyuk-A) [inline addons_marketplace/marketplace-manager/api/v1/marketplace_blueprint.py:1]: Do you have some exemplary SLAs for this to try out locally? Maybe even add a simple example SLA as a README?

File: addons_marketplace/marketplace-manager/Dockerfile
Comment (Malyuk-A) [inline addons_marketplace/marketplace-manager/Dockerfile:16]: image


```suggestion
CMD ["python", "marketplace_manager.py"]
```

File: addons_marketplace/marketplace-manager/marketplace_manager.py
Comment (Malyuk-A) [inline addons_marketplace/marketplace-manager/marketplace_manager.py:40]: Recommendation:

I am new to flask. During my own development I also stumbled upon this annoying warning:
image

I think to overcome this is pretty easy to do.

Have a look here how to use a "production ready WSGI server instead":
dependency needed: waitress = "3.0.0"

I use slightly different dependencies but the core thing can be resolved the same:

```python
from http import HTTPStatus

import flask_openapi3
import flops_manager.api.blueprints as blps
from flops_manager.api.utils.consts import FLOPS_MANAGER_PORT
from waitress import serve

_info = flask_openapi3.Info(title="FLOps Manager API", version="1.0.0")
app = flask_openapi3.OpenAPI(__name__, info=_info)


@app.route("/", methods=["GET"])
def health():
    return {"message": "ok"}, HTTPStatus.OK


def handle_api():
    for blp in blps.blueprints:
        app.register_api(blp)

    # Note (AFAIK): 'waitress' comes with its own logger.
    # We are already using the flops logger.
    # Thus we have duplicated logger outputs (not just for API things but generally).
    # To avoid this the waitress ones gets muted.
    # If something should be logged that concerns the API it will be logged via the flops logger.
    serve(app, host="0.0.0.0", port=FLOPS_MANAGER_PORT, _quiet=True)
```

File: addons_marketplace/marketplace-manager/db/mongodb_client.py
Comment (Malyuk-A) [inline addons_marketplace/marketplace-manager/db/mongodb_client.py:22]: How exactly does this work?
If I am not mistaken there is no new MongoDB container running, will python spawn this on its own?

File: addons_marketplace/marketplace-manager/services/marketplace_service.py
Comment (Malyuk-A) [inline addons_marketplace/marketplace-manager/services/marketplace_service.py:1]: Based on this diagram:

image

I expected to see a separate Addon Verifier component or even container. Is this your future vision of this and it is just for now a part of the marketplace-manager?

**Review by Malyuk-A — COMMENTED** [review]
Ok this is my initial review round for the marketplace part.
I will continue with the addons engine now.

File: addons_engine/addons-manager/services/addons_runner.py
Comment (Malyuk-A) [inline addons_engine/addons-manager/services/addons_runner.py:51]: nitpick:

"maybe_create_networks" sounds strange to me. How about "fix_unavailable_networks" or "try_to_fix_unavailable_networks"

File: addons_engine/addons-manager/services/addons_runner.py
Comment (Malyuk-A) [inline addons_engine/addons-manager/services/addons_runner.py:21]: Do you have some exemplary SLA/(test/dummy)addon I could try out to see if this works as intended?

File: addons_engine/addons-manager/tests/addons_service_test.py
Comment (Malyuk-A) [inline addons_engine/addons-manager/tests/addons_service_test.py:12]: If I am not mistaken:
```suggestion
from services.addons_runner import (  # noqa: E402
```

File: addons_engine/addons-manager/tests/addons_service_test.py
Comment (Malyuk-A) [inline addons_engine/addons-manager/tests/addons_service_test.py:1]: image

Could it be that this const is not defined?

image

File: addons_engine/addons-monitor/addons_monitor.py
Comment (Malyuk-A) [inline addons_engine/addons-monitor/addons_monitor.py:32]: Do you think it makes sense adding a sleep cmd here instead of just a pass? maybe this will be less "demanding" on the system?

**Review by Malyuk-A — COMMENTED** [review]
Completed my first review round for the addons_engine 🙂

File: addons_engine/addons-manager/db/addons_db.py
Comment (melkodary) [reply addons_engine/addons-manager/db/addons_db.py:11]: `find_addons` is function that is used, so just to increase reusability I believe this way it's better.

File: addons_marketplace/marketplace-manager/api/v1/schema.py
Comment (melkodary) [reply addons_marketplace/marketplace-manager/api/v1/schema.py:25]: similar to docker-compose, the services keyword indicates the containers.

File: addons_marketplace/marketplace-manager/services/marketplace_service.py
Comment (melkodary) [reply addons_marketplace/marketplace-manager/services/marketplace_service.py:13]: yes! inefficient but should be enough for now. I am working on a go implementation that does this more effieciently.

File: addons_marketplace/marketplace-manager/api/v1/marketplace_blueprint.py
Comment (melkodary) [reply addons_marketplace/marketplace-manager/api/v1/marketplace_blueprint.py:1]: Sure!

File: addons_marketplace/marketplace-manager/api/v1/marketplace_blueprint.py
Comment (Malyuk-A) [reply addons_marketplace/marketplace-manager/api/v1/marketplace_blueprint.py:22]: Ah yes I see. Ok makes sense

File: addons_marketplace/marketplace-manager/api/v1/schema.py
Comment (Malyuk-A) [reply addons_marketplace/marketplace-manager/api/v1/schema.py:25]: If you are fine with this that its ok. 
(For me who is not swimming in your topic I instantly associate this with oakestra services, so a simply change to addon_services would be cool, but up to you. I am fine with how it is now)

File: addons_marketplace/marketplace-manager/db/mongodb_client.py
Comment (melkodary) [reply addons_marketplace/marketplace-manager/db/mongodb_client.py:22]: you'd need to spin up the mongo container yourself if you want to use it, hence:
```
MONGO_URL = os.environ.get("ADDON_MARKETPLACE_MONGO_URL")
MONGO_PORT = os.environ.get("ADDON_MARKETPLACE_MONGO_PORT")
```

However I could make it more clear that for local development you'd need to spin up your own mongo db container.

File: addons_marketplace/marketplace-manager/services/marketplace_service.py
Comment (melkodary) [reply addons_marketplace/marketplace-manager/services/marketplace_service.py:1]: Thank you for pointing that out. . You are absolutely right and that's was my vision but for the time being I could indicate that it is just a process or something. I'll fix it.

File: addons_engine/start-up.sh
Comment (melkodary) [reply addons_engine/start-up.sh:1]: addon-engine cannot be run as a container, otherwise installed containers are docker in a docker, defeating it's purpose of running containers in the control plane.

File: addons_marketplace/marketplace-manager/db/mongodb_client.py
Comment (Malyuk-A) [reply addons_marketplace/marketplace-manager/db/mongodb_client.py:22]: This would be helpful, because all other oakestra things can be spun up somehow by compose files so this would reduce misconseptions/confusion

File: addons_engine/README.md
Comment (melkodary) [reply addons_engine/README.md:12]: Probably broke when I restructured some stuff. I'll fix it.

File: addons_engine/addons-manager/api/v1/addons_api.py
Comment (melkodary) [reply addons_engine/addons-manager/api/v1/addons_api.py:22]: same answer.

File: addons_engine/start-up.sh
Comment (Malyuk-A) [reply addons_engine/start-up.sh:1]: Isn't it possible to mount the docker.socket thus having access to the hosts docker engine and launch proper containers and not dind?

File: addons_engine/addons-manager/services/addons_runner.py
Comment (melkodary) [reply addons_engine/addons-manager/services/addons_runner.py:1]: I guess this could be cleaned up a little bit but 300 lines of code is a lot by which standard?

Comment (Malyuk-A) [reply addons_engine/addons-manager/services/addons_runner.py:1]: If you want to keep it as is ist is totally fine. I am simply used to python files that are < 120 LoCs 🙂

File: addons_engine/addons-manager/services/addons_runner.py
Comment (melkodary) [reply addons_engine/addons-manager/services/addons_runner.py:21]: Good point! I'll add some steps in the documentation for it.

File: addons_engine/addons-manager/tests/addons_service_test.py
Comment (melkodary) [reply addons_engine/addons-manager/tests/addons_service_test.py:12]: Oh yes! why does vscode not realize the refactor change 😠 Pycharm usually does.

File: addons_engine/addons-monitor/addons_monitor.py
Comment (melkodary) [reply addons_engine/addons-monitor/addons_monitor.py:32]: yeah I guess, it shouldn't loop forever I agree. I'll adjust.

File: addons_engine/addons-manager/tests/addons_service_test.py
Comment (Malyuk-A) [reply addons_engine/addons-manager/tests/addons_service_test.py:12]: Yeah, sometimes its a bit flaky 😅

File: addons_engine/start-up.sh
Comment (melkodary) [reply addons_engine/start-up.sh:1]: I have not used this before, but it seems it's mostly configuration stuff so its something we could implement later. However if this is possible I'd be wary of the security concerns this would have, since containers should be totally isolated and don't have access to the host outside.

Comment (Malyuk-A) [reply addons_engine/start-up.sh:1]: Fair points. Seems fine for now.

File: addons_marketplace/marketplace-manager/README.md
Comment (Malyuk-A) [inline addons_marketplace/marketplace-manager/README.md:16]:
```suggestion
$ docker run -p 11102:11102 marketplace_manager
```

The original cmd does not work for me

File: addons_engine/README.md
Comment (Malyuk-A) [inline addons_engine/README.md:12]: I get the same error:
image

Even if I install vevns like this: sudo apt install python3-virtualenv

File: addons_engine/addons-manager/tests/__init__.py
Comment (Malyuk-A) [inline addons_engine/addons-manager/tests/__init__.py:1]: The test works on my end now:
image

**Review by Malyuk-A — COMMENTED** [review]
I had a second look

File: addons_engine/README.md
Comment (melkodary) [reply addons_engine/README.md:12]: Do any start-up.sh in the project work for you?

Comment (melkodary) [reply addons_engine/README.md:12]: I quickly searched the error you reported. Is it possible you don't have python3.8 installed in your machine? If so try modifiying the start-up.sh line to be:
`virtualenv --clear -p python3 .venv` (python3 instead of python3.8)

File: addons_marketplace/marketplace-manager/services/marketplace_service.py
Comment (melkodary) [reply addons_marketplace/marketplace-manager/services/marketplace_service.py:32]: Currently I dont have a strong opinion about whether to keep it or remove failed ones. For the time being I dont see harm in keeping failed ones. Later if we find out that failed ones are no longer needed and they are taking a lot of space, we could setup a cron job the regularly cleans the db.

File: addons_engine/README.md
Comment (Malyuk-A) [reply addons_engine/README.md:12]: It works when I change pyhton3.8 to python3.
It would be cool if you could change that so that these scripts are not hardcoded to version 3.8 (I am using 3.10 and up)

File: addons_marketplace/marketplace-manager/services/marketplace_service.py
Comment (giobart) [inline addons_marketplace/marketplace-manager/services/marketplace_service.py:17]: If I'm not mistaken from the docker python SDK there is a command method `get_registry_data` that gives you the metadata from the image registry given an image name. No image pull requested. Maybe we can use that for this check?

**Review by giobart — CHANGES_REQUESTED** [review]

File: addons_engine/addons-manager/addons_manager.py
Comment (melkodary) [reply addons_engine/addons-manager/addons_manager.py:22]: I am modifying the logging configuration for testing purposes, making it printout everything even info.  So i just added a TODO for me to remove it before merging

File: addons_engine/addons-manager/services/addons_service.py
Comment (melkodary) [reply addons_engine/addons-manager/services/addons_service.py:18]: What I have in mind is that the user can issue either disable or install instructions, and when actually the addon is not running anymore because the `addons_monitor` stopped it, the status shown will be 'disabled'. Same case for the install. So when it's actually running it will be changed to active. 

I would love some feedback here because I have been having doubts about the different statuses and I believe it is still confusing somehow.

File: addons_engine/addons-monitor/utils/cleanup_handler.py
Comment (melkodary) [reply addons_engine/addons-monitor/utils/cleanup_handler.py:5]: The addons_monitor.py when it is being shutdown. This is called when a SIGINT or SIGTERM is triggered.
```
def signal_handler(sig, frame):
    logging.info("Shutting down Addon Manager...")
    handle_shutdown()

    sys.exit(0)


if __name__ == "__main__":
    signal.signal(signal.SIGINT, signal_handler)
    signal.signal(signal.SIGTERM, signal_handler)
    try:
        addons_monitor.start_monitoring()  # This is a blocking call
    except Exception as e:
        logging.error("An error occurred while monitoring addons", exc_info=e)
```

File: addons_marketplace/marketplace-manager/services/marketplace_service.py
Comment (melkodary) [reply addons_marketplace/marketplace-manager/services/marketplace_service.py:17]: Oh I see, that's even better. Thanks a lot :D

File: addons_engine/addons-manager/services/addons_service.py
Comment (melkodary) [reply addons_engine/addons-manager/services/addons_service.py:18]: I could change it to, `installing` & `disabling`, indicating that it is not yet disabled or active.

File: addons_engine/addons-monitor/utils/cleanup_handler.py
Comment (melkodary) [reply addons_engine/addons-monitor/utils/cleanup_handler.py:5]: Just realized that this doesn't work when the process is running as a container.

**Comment by giobart** [comment]
Can we migrate this implementation as compose override instead of directly in root compose? @melkodary

**Review by giobart — APPROVED** [review]

**Pull request merged**
